lib/winapi/xapi.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2019 Stefan Schmidt
+ *
+ * Licensed under the MIT License
+ */
+
+#include <xapi.h>
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <xboxkrnl/xboxkrnl.h>
+
+__attribute__((constructor)) static VOID setup_d_drive (VOID)
+{
+    NTSTATUS status;
+    HANDLE handle;
+    ANSI_STRING driveString = RTL_CONSTANT_STRING("\\??\\D:");
+    OBJECT_ATTRIBUTES objattr;
+
+    InitializeObjectAttributes(&objattr, &driveString, OBJ_CASE_INSENSITIVE, NULL, NULL);
+    status = NtOpenSymbolicLinkObject(&handle, &objattr);
+    if (NT_SUCCESS(status)) {
+        NtClose(handle);
+        return;
+    }
+
+    // D: doesn't exist yet, so we create it
+    CHAR targetPath[MAX_PATH];
+
+    // Retrieve the XBE path
+    strncpy(targetPath, XeImageFileName->Buffer, XeImageFileName->Length);
+    targetPath[XeImageFileName->Length] = '\0';
+
+    // Cut off the XBE file name by inserting a null-terminator
+    PCHAR filenameStr;
+    filenameStr = strrchr(targetPath, '\\');
+    assert(filenameStr != NULL);
+    *(filenameStr + 1) = '\0';
+
+    // Mount the remaining path as D:
+    BOOL success;
+    success = XMountDrive('D', targetPath);
+    assert(success);
+}
+
+BOOL XMountDrive (CHAR driveLetter, PCSTR path)
+{
+    NTSTATUS status;
+    ANSI_STRING drivePath;
+    CHAR drivePathBuffer[7];
+    ANSI_STRING mountPath;
+
+    sprintf(drivePathBuffer, "\\??\\%c:", driveLetter);
+    RtlInitAnsiString(&drivePath, drivePathBuffer);
+
+    RtlInitAnsiString(&mountPath, path);
+
+    status = IoCreateSymbolicLink(&drivePath, &mountPath);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+BOOL XUnMountDrive (CHAR driveLetter)
+{
+    NTSTATUS status;
+    ANSI_STRING drivePath;
+    CHAR drivePathBuffer[7];
+
+    sprintf(drivePathBuffer, "\\??\\%c:", driveLetter);
+    RtlInitAnsiString(&drivePath, drivePathBuffer);
+
+    status = IoDeleteSymbolicLink(&drivePath);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+BOOL XReMountDrive (CHAR driveLetter, PCSTR path)
+{
+    XUnMountDrive(driveLetter);
+    return XMountDrive(driveLetter, path);
+}
```

Comment:
I don't think this function is necessary? If the path is already mounted, then the user should probably look into dealing with it (why was it mounted? maybe it had a purpose?).

I think a more suitable API design would possibly be to be able to `bool force` a `XMountDrive`.

Eitherway, I see drawbacks when unmounting succeeds, but remounting fails.

---

Realisticly, I assume that we'll provide some optional XDK XAPI-like lib which respects XBE flags and does the XDK-like mounting at startup. Most applications would probably use that. So it's unlikely the user would want to manually mount something (and if they do, they are probably fine with more steps).

Reply:
Removed.

> So it's unlikely the user would want to manually re-mount something (and if they do, they are probably fine with more steps).
If we're talking about games or something similar here, I agree. But I think it's quite likely that utility programs want to access the full partitions instead of only game-specific paths.

Reply:
I was specifically thinking of something like a file-manager. Those utilities would want to ensure that the paths they re-mounted are mounted properly (as they don't want users to accidentally manipulate the wrong filesystem). So they are fine with extra-steps and more careful error-checking.